sw/host/tests/usbdev/usbdev_stream/usbdev_stream.cc

```diff
@@ -0,0 +1,391 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+#include "usbdev_stream.h"
+
+#include <cassert>
+#include <cstdio>
+#include <cstring>
+#include <iostream>
+
+#include "stream_test.h"
+#include "usb_device.h"
+#include "usbdev_utils.h"
+
+// Stream signature words.
+#define STREAM_SIGNATURE_HEAD 0x579EA01AU
+#define STREAM_SIGNATURE_TAIL 0x160AE975U
+
+// Seed numbers for the LFSR generators in each transfer direction for
+// the given stream number.
+#define USBTST_LFSR_SEED(s) (uint8_t)(0x10U + (s)*7U)
+#define USBDPI_LFSR_SEED(s) (uint8_t)(0x9BU - (s)*7U)
+
+// Simple LFSR for 8-bit sequences.
+#define LFSR_ADVANCE(lfsr) \
+  (((lfsr) << 1) ^         \
+   ((((lfsr) >> 1) ^ ((lfsr) >> 2) ^ ((lfsr) >> 3) ^ ((lfsr) >> 7)) & 1u))
+
+USBDevStream::USBDevStream(unsigned id, uint32_t transfer_bytes, bool retrieve,
+                           bool check, bool send, bool verbose) {
+  // Remember Stream IDentifier and flags.
+  SetProperties(id, retrieve, check, send);
+  verbose_ = verbose;
+
+  // Stream is not in the process of being closed.
+  closing_ = false;
+
+  // Not yet received stream signature.
+  sig_recvd_ = kSigStateStart;
+
+  // Initialise LFSR state.
+  tst_lfsr_ = USBTST_LFSR_SEED(id);
+  dpi_lfsr_ = USBDPI_LFSR_SEED(id);
+
+  // Initialize circular buffer.
+  buf_.wr_idx = 0U;
+  buf_.rd_idx = 0U;
+  buf_.end_idx = 0U;
+
+  // Number of bytes to be transferred.
+  transfer_bytes_ = transfer_bytes;
+
+  // Total counts of bytes received and sent.
+  bytes_recvd_ = 0U;
+  bytes_sent_ = 0U;
+}
+
+// Detect a stream signature within the byte stream;
+// simple restartable parser that drops all bytes until we find a header
+// signature.
+//
+// The signature permits usbdev_stream_test to pass test parameters to this
+// program as well as overcoming the issue of the mapping from device
+// endpoints/streams to USB ports not being under our control.
+uint32_t USBDevStream::SigDetect(usbdev_stream_sig_t *sig, const uint8_t *sp,
```

Comment:
I find this function to be quite confusing and hard to follow. Would the following work:
- accumulate bytes in `sig` until you have `sizeof(usbdev_stream_sig_t)` bytes
- at this point, verify if this is a valid signature, if so then great
- if not, drop the first byte, shift everyelse one byte

The only downside it would have compared to yours is that it takes more bytes before you starts discarding. But I think it would also make it much easier to follow.